OCI runtime support for nspawn

poettering opened the pull request:
This is a big one, unfortunately.

For full drop-in OCI runtime support two bits are still missing: the OCI hooks need to be executed at the right times, and we need a command line tool that provides runc compatibility. I am on both, but this already grew large enough I didn't want to pile on.

But even without those two bits it's pretty comprehensive and should pretty much work.

poettering commented:
/cc @alban, @iaguis, @dongsupark

boucman commented on man/systemd-nspawn.xml:
```diff
@@ -230,6 +230,15 @@
+
+        <listitem><para>Takes the path to an OCI runtime bundle to invoke, as specified in the <ulink
+        url="https://github.com/opencontainers/runtime-spec/blob/master/spec.md">OCI Runtime Specification</ulink>. In
+        this case no <filename>.settings</filename> file is loaded, and the root directory and various settings are
```
you mean .nspawn file here ?

poettering replied:
i do!

evverx commented:
This pull request **introduces 6 alerts** when merging a9451bae2a277b0aa4889fd3d1badf335ed28a89 into 5a8b16409240bc95d95f3c19d6f5cd9366e6ccf4 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-1a21c6a841a0fbce8ed251a82654edf1be872603)

**new alerts:**

* 6 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

evverx commented:
This pull request **introduces 6 alerts** when merging 7572186e36e7553a6f8c6bee3ff51c20ef97d80b into 5a8b16409240bc95d95f3c19d6f5cd9366e6ccf4 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-26db5d9a33529e3cdf426ad77d32a7bcc23c9258)

**new alerts:**

* 6 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

dnicolodi commented on src/basic/json-internal.h:
```diff
@@ -0,0 +1,52 @@
+#define JSON_VALUE_NULL ((JsonValue) {})
+
+/* We use fake JsonVariant objects for some special values, in order to avoid memory allocations for them. Note that
+ * effectively this means that there are multiple ways to encode the some objects: via these magic values or as
```
`encode the some objects` -> `encode some objects`

dnicolodi replied:
Typo still here :)

evverx commented on src/basic/json.h:
```diff
@@ -0,0 +1,272 @@
+
+/*
+  In case you wonder why we have our own JSON implementation, here are a couple of reasons why this implementation has
+  benefits over various other implementatins:
```
I'm not going to say it's a bold move to implement a json parser :-), but if it's going to be implemented, maybe it would be helpful to start fuzzing it right away. I ran a fuzzer  compiled with UBSan for about ten seconds and it crashed after receiving relatively normal input like `-34444444444541444444` and `"0"` with
```sh
../../src/systemd/src/basic/json.c:1845:48: runtime error: signed integer overflow: 10 * -3444444444454144444 cannot be represented in type 'long'
    #0 0x7fe6beefb805 in json_parse_number /work/build/../../src/systemd/src/basic/json.c:1845:48
    #1 0x7fe6beefa5ca in json_tokenize /work/build/../../src/systemd/src/basic/json.c:2009:29
```
and
```
../../src/systemd/src/basic/json.c:350:9: runtime error: index 1 out of bounds for type 'char [0]'
    #0 0x7fab2d4c2042 in json_variant_new_stringn /work/build/../../src/systemd/src/basic/json.c:350:22
    #1 0x7fab2d4c8690 in json_parse_internal /work/build/../../src/systemd/src/basic/json.c:2321:29
    #2 0x7fab2d4c781a in json_parse /work/build/../../src/systemd/src/basic/json.c:2477:16
```
There were also a couple warnings from `clang`, the first of which seems legit:
```
../../src/systemd/src/basic/json.c:2486:32: warning: unused variable 'opened' [-Wunused-variable]
        _cleanup_fclose_ FILE *opened = NULL;
                               ^
../../src/systemd/src/basic/json.c:2949:13: warning: variable 'source' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
        if (variant) {
            ^~~~~~~
../../src/systemd/src/basic/json.c:2955:13: note: uninitialized use occurs here
        if (source && source_line > 0 && source_column > 0)
            ^~~~~~
../../src/systemd/src/basic/json.c:2949:9: note: remove the 'if' if its condition is always true
        if (variant) {
        ^~~~~~~~~~~~~
../../src/systemd/src/basic/json.c:2936:27: note: initialize the variable 'source' to silence this warning
        const char *source;
                          ^
                           = NULL
```

evverx replied:
Unfortunately, false positives are inevitable, but I don't think they make fuzzers less useful. In this particular case, `__attribute__((no_sanitize("signed-integer-overflow")))` and `__attribute__((no_sanitize("bounds")))` could probably be used to silence UBSan.

poettering replied:
I figure I could rewrite the overflow check to not trigger ubsan. But for the [0] bounds check I have no idea how I could trick it to not complain. Any idea?

Quite frankly, ubsan should be smart enough to understand that [0] is special, and generally doesn't actually mean "zero sized array", but instead is similar to C99 [], i.e. more akin to "unspecified size"...

evverx replied:
I've done some research and it turns out that UBSan is smart enough to understand that [0] is special as long as it's the last element of a union within a structure. Placing `char string[0]` at the end of the union will make the warning go away.

evverx replied:
BTW I've written a fuzzer. It's waiting to be unleashed in https://github.com/systemd/systemd/pull/9782 :-)

poettering replied:
@evverx hmm, I figure it would make sense to simply pick up that #9782 and add it to this PR, what do you think?

evverx replied:
@poettering it's fine by me, but it shouldn't be merged into master until https://github.com/google/oss-fuzz/pull/1683 is merged.

evverx replied:
@poettering https://github.com/google/oss-fuzz/pull/1683 has been merged. I pushed the commit from #9782 on top of your branch.

evverx replied:
Fedora CI seems to have vanished. Does anybody know how to bring it back or trigger a build manually?

evverx replied:
I'm not sure what's happened, but the fuzzer is gone. It'd probably be better to leave it util later. I'll reopen #9782 so as not to forget that it isn't done yet.

evverx commented on src/busctl/busctl.c:
```diff
@@ -1656,7 +2029,21 @@ static int get_property(int argc, char **argv, void *userdata) {
+
+                        json_dump_with_flags(v, stdout);
+                        if (r < 0)
+                                return r;
```
It'd be great if you could also take a look at the LGTM alerts. At least the last two of them don't look like false positives. For example, the if block really seems to be redundant here.

rhatdan commented:
@giuseppe FYI

poettering commented:
I force pushed a new version now, addressing all raised issues. Let's see if LGTM likes it more this time. PTAL.

evverx commented:
`test-stat-util` seems to have started to fail on Fedora CI with
```
152/321 test-stat-util                          FAIL     0.19 s (killed by signal 6 SIGABRT)
--- command ---
RPM_OS='linux' _='/usr/bin/python3' LANG='C' container_uuid='05c04656-05d9-479d-aa1d-84652cc6678d' HISTCONTROL='ignoredups' RPM_ARCH='x86_64' HOSTNAME='' OLDPWD='/builddir/build/BUILD' CONFIG_SITE='NONE' RPM_PACKAGE_RELEASE='0.1.20180802134457.470418a.fc29' RPM_BUILD_DIR='/builddir/build/BUILD' RPM_LD_FLAGS='-Wl,-z,relro   -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld' NOTIFY_SOCKET='/run/systemd/nspawn/notify' container='systemd-nspawn' USER='mockbuild' RPM_SOURCE_DIR='/builddir/build/SOURCES' PWD='/builddir/build/BUILD/systemd-240/x86_64-redhat-linux-gnu' RPM_PACKAGE_VERSION='240' HOME='/builddir' RPM_BUILD_ROOT='/builddir/build/BUILDROOT/systemd-240-0.1.20180802134457.470418a.fc29.x86_64' MAIL='/var/spool/mail/mockbuild' SHELL='/bin/bash' TERM='vt100' RPM_PACKAGE_NAME='systemd' RPM_DOC_DIR='/usr/share/doc' SHLVL='3' PROMPT_COMMAND='printf "\033]0;<mock-chroot>\007"' LOGNAME='mockbuild' PATH='/builddir/build/BUILD/systemd-240/x86_64-redhat-linux-gnu:/builddir/.local/bin:/builddir/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin' PKG_CONFIG_PATH=':/usr/lib64/pkgconfig:/usr/share/pkgconfig' RPM_OPT_FLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection' HISTSIZE='1000' LC_CTYPE='C.UTF-8' SYSTEMD_KBD_MODEL_MAP='/builddir/build/BUILD/systemd-240/src/locale/kbd-model-map' SYSTEMD_LANGUAGE_FALLBACK_MAP='/builddir/build/BUILD/systemd-240/src/locale/language-fallback-map' /builddir/build/BUILD/systemd-240/x86_64-redhat-linux-gnu/test-stat-util
--- stderr ---
Assertion 'device_path_make_canonical(st.st_mode, st.st_rdev, &resolved) >= 0' failed at ../src/test/test-stat-util.c:113, function test_device_path_make_canonical_one(). Aborting.
```
It probably has something to do with the absence of `/run/systemd/inaccessible`, but I wouldn't say I'm 100% certain.

poettering commented:
Neat! CI all green now!

evverx commented:
Interestingly, it seems that LGTM restarted the analysis as soon as the label was removed.

poettering commented:
I force pushed another new version now, fixing the typo @dnicolodi found.

keszybz commented:
Hmmm, can we get the non-json parts as a separate PR?

keszybz commented:
This gives a bunch of warnings during build:
```
../src/basic/macro.h:231:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
                 UNIQ_T(A, aq) > UNIQ_T(B, bq) ? 1 : 0;  \
                               ^
../src/basic/macro.h:26:44: note: in definition of macro ‘_unlikely_’
 #define _unlikely_(x) (__builtin_expect(!!(x), 0))
                                            ^
../src/basic/macro.h:301:25: note: in expansion of macro ‘assert_message_se’
 #define assert_se(expr) assert_message_se(expr, #expr)
                         ^~~~~~~~~~~~~~~~~
../src/test/test-util.c:97:9: note: in expansion of macro ‘assert_se’
         assert_se(CMP(0L, UINT64_MAX) == -1);
         ^~~~~~~~~
../src/basic/macro.h:225:19: note: in expansion of macro ‘__CMP’
 #define CMP(a, b) __CMP(UNIQ, (a), UNIQ, (b))
                   ^~~~~
../src/test/test-util.c:97:19: note: in expansion of macro ‘CMP’
         assert_se(CMP(0L, UINT64_MAX) == -1);
                   ^~~
```

keszybz commented:
Crashes with with attached test case.
```==11==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6060005eb420 at pc 0x7fddb6ac0c58 bp 0x7ffdf75374f0 sp 0x7ffdf75374e8
WRITE of size 1 at 0x6060005eb420 thread T0
SCARINESS: 31 (1-byte-write-heap-buffer-overflow)
    #0 0x7fddb6ac0c57 in json_parse_string /work/build/../../src/systemd/src/basic/json.c:1787:38
    #1 0x7fddb6abf835 in json_tokenize /work/build/../../src/systemd/src/basic/json.c:2065:29
    #2 0x7fddb6ac1fa1 in json_parse_internal /work/build/../../src/systemd/src/basic/json.c:2239:25
    #3 0x7fddb6ac4426 in json_parse_file /work/build/../../src/systemd/src/basic/json.c:2572:16
    #4 0x5303cb in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/fuzz/fuzz-json.c:20:13
    #5 0x5684a8 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:576:15
    #6 0x566317 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/libfuzzer/FuzzerLoop.cpp:485:3
    #7 0x56a272 in fuzzer::Fuzzer::MutateAndTestOne() /src/libfuzzer/FuzzerLoop.cpp:713:19
    #8 0x56d206 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:844:5
    #9 0x54cc74 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:765:6
    #10 0x5307ac in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #11 0x7fddb583382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #12 0x41cff8 in _start (/out/fuzz-json+0x41cff8)

0x6060005eb420 is located 0 bytes to the right of 64-byte region [0x6060005eb3e0,0x6060005eb420)
allocated by thread T0 here:
    #0 0x4ec19e in realloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:165
    #1 0x7fddb6a3198b in greedy_realloc /work/build/../../src/systemd/src/basic/alloc-util.c:55:13
    #2 0x7fddb6ac074f in json_parse_string /work/build/../../src/systemd/src/basic/json.c:1825:38
    #3 0x7fddb6abf835 in json_tokenize /work/build/../../src/systemd/src/basic/json.c:2065:29
    #4 0x7fddb6ac1fa1 in json_parse_internal /work/build/../../src/systemd/src/basic/json.c:2239:25
    #5 0x7fddb6ac4426 in json_parse_file /work/build/../../src/systemd/src/basic/json.c:2572:16
    #6 0x5303cb in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/fuzz/fuzz-json.c:20:13
    #7 0x5684a8 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:576:15
    #8 0x566317 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/libfuzzer/FuzzerLoop.cpp:485:3
    #9 0x56a272 in fuzzer::Fuzzer::MutateAndTestOne() /src/libfuzzer/FuzzerLoop.cpp:713:19
    #10 0x56d206 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:844:5
    #11 0x54cc74 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:765:6
    #12 0x5307ac in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #13 0x7fddb583382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow /work/build/../../src/systemd/src/basic/json.c:1787:38 in json_parse_string
Shadow bytes around the buggy address:
  0x0c0c800b5630: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b5640: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b5650: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b5660: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b5670: fa fa fa fa fa fa fa fa fa fa fa fa 00 00 00 00
=>0x0c0c800b5680: 00 00 00 00[fa]fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b5690: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b56a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b56b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b56c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c800b56d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==11==ABORTING
```

keszybz commented:
http://in.waw.pl/~zbyszek/systemd//crash-cf07f1aa943d77db87163ecbc3c55e67f31c07bf

filbranden commented:
> This gives a bunch of warnings during build:
> (`...: comparison between signed and unsigned integer expressions ...`)

These are fixed by commit ef6e83f02b017de94204faeaf2a2cdf3f4c868a2.

vbatts commented:
cc @vbatts ;-)

boucman commented on man/journalctl.xml:
```diff
@@ -316,10 +316,23 @@
+                information). Field values are generally encoded as JSON strings, with three exceptions:
+                <orderedlist>
+                  <listitem><para>Fields larger than 4096 bytes are encoded as <constant>null</constant> values. (This
+                  may be turned off by passing <option>--all</option>, but be aware that this my allocate overly long
```
this MAY allocate

boucman commented on man/systemd-nspawn.xml:
```diff
@@ -1060,6 +1078,42 @@
+        <option>interactive</option>, <option>read-only</option>, <option>passive</option> or <option>pipe</option>. If
+        <option>interactive</option> a pseudo-TTY is allocated and made available as <filename>/dev/console</filename>
+        in the container. It is then continously bi-directionally connected to the standard input and output passed by
+        the calling process.  <option>read-only</option> is similar but only the output of the container is propagated
```
in this case, the calling process is always systemd-nspawn, right ? if yes, it might be clearer to say that explicitely...

poettering replied:
no, it's the caller of nspawn itself that is meant here. The man page is generally written from perspective of nspawn. I'll slightly reword this to make this clearer.

keszybz commented:
> can we get the non-json parts as a separate PR?

That would be nice. This PR needs a rebase now anyway.

poettering commented:
I force pushed a new version now, with everything reported fixed. I also split out various bits into a number of other PRs. The JSON stuff is now in #10316, and should really be reviewed first (the main json patch is also included in this PR to make sure it builds fine though, will rebase as soon as #10316 is merged).

The most relevant other PRs split out from this one are #10317 and #10318.

PTAL!

poettering commented:
(and yeah, the memory corruption @keszybz reported is fixed in #10316 too)

dongsupark commented:
First of all, I have to say, I like this feature. :+1: 
I have not managed to read all the code, but in general it looks good to me.

Though I have a question.
When running simply `systemd-nspawn --oci-bundle=/path/to/container`, it fails because it cannot find a rootfs. So I had to specify an additional option `--directory=/path/to/container`, to make it work. It looks a little strange, because an oci bundle is supposed to already include the rootfs.
Is it intended?

rhatdan commented:
Has anyone attempted this with a tool like podman?  Supporting reading json is one thing, but there are also some CLI commands that need to match up with runc, to make different container runtimes plugable with container engines like CRI-O, Docker, Podman, Buildah.

rhatdan commented:
I guess what I am suggesting is we have a systemd-nspawn-oci as a symlink of systemd-nspawn and that systemd-nspawn have a CLI that matches runc CLI.  Then it would be simple to do

podman --runtime /usr/bin/systemd-nspawn-oci run fedora ...

ANd it would use systemd-nspawn instead of runc.  

Docker, CRI-O and buildah all have similar ways of replacing the OCI Runtime, but will launch them all with the `runc` default CLI.

poettering commented:
> When running simply `systemd-nspawn --oci-bundle=/path/to/container`, it fails because it cannot find a rootfs. So I had to specify an additional option `--directory=/path/to/container`, to make it work. It looks a little strange, because an oci bundle is supposed to already include the rootfs.
> Is it intended?

No this is really supposed to work without --directory=. Does your JSON file contain the rootfs reference or is it omitted?

Can you supply me with a tarball of the tree you used this one where this failed?

poettering commented:
> I guess what I am suggesting is we have a systemd-nspawn-oci as a symlink of systemd-nspawn and that systemd-nspawn have a CLI that matches runc CLI.

Yes, it is definitely my intention to make this work. Alas the CLI is a de-facto standard only, so I need to reverse engineer the minimal command set I actually need to support, and what these things actually really do and what things rely on. For example, iiuc runc sometimes forks off its stuff into the background (what is this, 1995 UNIX???) and then callers of runc use PR_SET_SUBREAPER to get notifications when that stuff forked into the background has exited. These semantics are just awful, and emulating them is nasty, since I'd very much prefer not to have to fork stuff into the background like that (instead, i want to do it properly, and run things as proper systemd services if background operation is requested), just to make runc callers happy which expect that happens.

So, before I can supply the CLI I need to do some research what the various tools actually want from runc, and what the semantics are they expect.

rhatdan commented:
I don't see you having to emulate that.  @giuseppe has integrated podman to use gvisord and we have cri-o using Kata Containers.   I would simply try to get podman to run with systemd-oci as CLI Emulation mode should be enough.

flx42 commented:
@rhatdan does CRI-O requires OCI runtimes to implement the following [operations](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md#operations) at the CLI level? Or is there more to it?

rhatdan commented:
@flx42 We need those commands.  As they are standardized.  But a few of the runc options have become defacto standard, just because the container engine's have built in calls so some of these options, based on using runc.

keszybz commented on src/basic/path-util.c:
```diff
@@ -495,6 +495,70 @@ char* path_join(const char *root, const char *path, const char *rest) {
+
+        /* Joins all listed strings until NULL and places an "/" between them unless the strings end/begin already with
+         * one so that it is unnecessary. Note that "/" which are already duplicate won't be removed. The string
+         * returned is hence always longer than the sum of the length of each individual string.
```
The second example in the list below is a counterexample.

Also "length**s**".

keszybz commented on src/basic/path-util.h:
```diff
@@ -49,6 +49,7 @@ int path_compare(const char *a, const char *b) _pure_;
 bool path_equal(const char *a, const char *b) _pure_;
 bool path_equal_or_files_same(const char *a, const char *b, int flags);
 char* path_join(const char *root, const char *path, const char *rest);
+char* path_join_many(const char *first, ...);
```
I'd prefer to have this wrapped like `strjoina` and others (with the real function renamed and `#define path_join_many(x, ...) _path_join_many(x, __varargs__, NULL)`.

keszybz commented on src/basic/stat-util.c:
```diff
@@ -266,3 +268,99 @@ int fd_verify_regular(int fd) {
+                return -ENODEV;
+
+        if (asprintf(ret, "/dev/%s/%u:%u", t, major(devno), minor(devno)) < 0)
+                return -ENOMEM;
```
asprintf(3) says that `ret` is undefined on failure. So I think this should go through a level of inderection where `ret` is assigned only on success, and the pointer is freed if non-zero on failure internally here. This way the callers don't have to care about the implementation details on failure.

poettering replied:
we had this discussion in some github comments before. my reading of the man page is that it is undefined as in "not overwritten", i.e. set to whatever it was set before, whatever that might be. hence it's not really undefined, just as undefined as what was in there before.

This function is a GNU function and glibc does not override it on failure (why would it even?), hence I think we should be fine on this mostly, at least until they change their code on this.

Or to say this different: if we really have a problem here, then we need to fix things all over the place, and when we do that I think the better approach is to add asprintf_safe() or so, that gives our usual "no-clobber-on-failure" guarantee explicitly... and then patch our entire codebase for it.

But quite frankly, for now I think we don't really need to bother...

keszybz replied:
I read "undefined" as meaning "either not set at all, or set to a valid pointer returned by malloc()". In other words, it's undefined whether the allocation occurred, but if anything is written to the pointer, it is a valid pointer. In the codebase, we always initialize the pointer to NULL and call free() on it, so we handle both possibilities correctly.

poettering commented:
force pushed a new version. only changes are the path_join_may() fixes @keszybz pointed out. asprintf() usage i left the way it is, for the mentioned reasons

poettering commented:
I force pushed a new version now, with no changes but rebase now that #10306 has been merged.

keszybz commented:
Something seems to be off with line number reporting:
```json
{
    "ociVersion": "1.0.0",

    "linux": {
	"capabilities": [
	    "CAP_AUDIT_WRITE",
	    "CAP_KILL",
	    "CAP_NET_BIND_SERVICE"
	]
    }
}
```
```console
$ sudo SYSTEMD_LOG_LEVEL=debug ~/src/systemd-work2/build/systemd-nspawn --oci-bundle=.
...
/home/zbyszek/src/oci-runtime-tests/basic/./config.json:75: Unexpected OCI element 'capabilities' of type 'array'.
```

keszybz commented:
When I'm running this on a host systemd with cgroups v2, it fails:
```
(sd-namespace) succeeded.
Init process invoked as PID 4909
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.85405 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.machine1 path=/org/freedesktop/machine1 interface=org.freedesktop.machine1.Manager member=CreateMachineWithNetwork cookie=3 reply_cookie=0 signature=sayssusaia(sv) error-name=n/a error-message=n/a
Got message type=method_return sender=:1.3 destination=:1.85405 path=n/a interface=n/a member=n/a cookie=365 reply_cookie=3 signature=o error-name=n/a error-message=n/a
Mounting cgroup on /tmp/unifiedoRljV3 (MS_NOSUID|MS_NODEV|MS_NOEXEC "none,name=systemd,xattr")...
Failed to chown "/tmp/unifiedoRljV3/machine.slice/machine-rootfs.scope/cgroup.controllers", ignoring: No such file or directory
Failed to chown "/tmp/unifiedoRljV3/machine.slice/machine-rootfs.scope/cgroup.events", ignoring: No such file or directory
Failed to chown "/tmp/unifiedoRljV3/machine.slice/machine-rootfs.scope/cgroup.stat", ignoring: No such file or directory
Failed to chown "/tmp/unifiedoRljV3/machine.slice/machine-rootfs.scope/cgroup.subtree_control", ignoring: No such file or directory
Failed to chown "/tmp/unifiedoRljV3/machine.slice/machine-rootfs.scope/cgroup.threads", ignoring: No such file or directory
Umounting /tmp/unifiedoRljV3...
Failed to enable controller cpu for /machine.slice/machine-rootfs.scope (/sys/fs/cgroup/machine.slice/machine-rootfs.scope/cgroup.subtree_control): No such file or directory
Failed to chown "/sys/fs/cgroup/machine.slice/machine-rootfs.scope/payload/cgroup.clone_children", ignoring: No such file or directory
Failed to chown "/sys/fs/cgroup/machine.slice/machine-rootfs.scope/payload/notify_on_release", ignoring: No such file or directory
Failed to chown "/sys/fs/cgroup/machine.slice/machine-rootfs.scope/payload/tasks", ignoring: No such file or directory
Failed to chown "/sys/fs/cgroup/machine.slice/machine-rootfs.scope/payload/cgroup.clone_children", ignoring: No such file or directory
Failed to chown "/sys/fs/cgroup/machine.slice/machine-rootfs.scope/payload/notify_on_release", ignoring: No such file or directory
Failed to chown "/sys/fs/cgroup/machine.slice/machine-rootfs.scope/payload/tasks", ignoring: No such file or directory
Bind-mounting /sys/fs/cgroup/systemd/machine.slice/machine-rootfs.scope/payload on /sys/fs/cgroup/systemd/machine.slice/machine-rootfs.scope/payload (MS_BIND "")...
Failed to mount /sys/fs/cgroup/systemd/machine.slice/machine-rootfs.scope/payload (type n/a) on /sys/fs/cgroup/systemd/machine.slice/machine-rootfs.scope/payload (MS_BIND ""): No such file or directory
Child died too early.
Bus n/a: changing state RUNNING → CLOSED
```

keszybz commented:
With (taken from the OCI spec example):
```json
       "capabilities": {
            "bounding": [
                "CAP_AUDIT_WRITE",
                "CAP_KILL",
                "CAP_NET_BIND_SERVICE"
            ],
            "permitted": [
                "CAP_AUDIT_WRITE",
                "CAP_KILL",
                "CAP_NET_BIND_SERVICE"
            ],
            "inheritable": [
                "CAP_AUDIT_WRITE",
                "CAP_KILL",
                "CAP_NET_BIND_SERVICE"
            ],
            "effective": [
                "CAP_AUDIT_WRITE",
                "CAP_KILL"
            ],
            "ambient": [
                "CAP_NET_BIND_SERVICE"
            ]
    }
```
it fails with
```
drop_capabilities() failed: Operation not permitted
Child died too early.
```

poettering commented:
I have now rebased evertyhing, and addressed a number of issues pointed out. I also split out everything that is not the OCIfication itself into a separate PR #10996, let's get that merged, to minimize my rebase burden. @keszybz ptal

poettering commented:
now that #10996 got merged, rebased, so that this is purely the OCI patch now

poettering commented on src/nspawn/nspawn-oci.c:
```diff
@@ -0,0 +1,2350 @@
+static int oci_terminal(const char *name, JsonVariant *v, JsonDispatchFlags flags, void *userdata) {
+        Settings *s = userdata;
+
+        s->console_mode = json_variant_boolean(v) ? CONSOLE_PASSIVE : CONSOLE_PIPE;
```
hmm, the OCI spec is extremely vague about all this. I have now changed ehaviour here though if terminal is set to true we'll now just do the normal behaviour with the console: i.e. interactive when we are invoked on a tty, and read-only on everything else. This means that only terminal=false has any effec tnow, as that turns off the default behaviour and forces us into --pipe mode.

poettering commented:
> When I'm running this on a host systemd with cgroups v2, it fails:

THis is actually a bug in nspawn independent of OCI. Will file bugfix independently of this shortly.

poettering commented:
> Something seems to be off with line number reporting:

This is also a bug independent of this PR. Will post fix shortly

poettering commented on src/nspawn/nspawn-oci.c:
```diff
@@ -0,0 +1,2350 @@
+                { "noNewPrivileges", JSON_VARIANT_BOOLEAN, json_dispatch_boolean, offsetof(Settings, no_new_privileges), 0               },
+                { "oomScoreAdj",     JSON_VARIANT_INTEGER, oci_oom_score_adj,     0,                                     0               },
+                { "selinuxLabel",    JSON_VARIANT_STRING,  oci_unsupported,       0,                                     JSON_PERMISSIVE },
+                { "user",            JSON_VARIANT_OBJECT,  oci_user,              0,                                     0               },
```
the spec is pretty clear about this and says the uid and gid fields are both mandatory, hence I will change the code to also require it.

poettering replied:
If I put "uid": 1.3, there's no error reported... I think it should fail. this is actually an artifact from the fact that it previously wasn't mandatory. will fix this too.

poettering replied:
> When I do manage to get a user, I see HOME=/root, SHELL=/sbin/nologin, LOGNAME=root. I'm not sure what is setting those, but they are bogus.

this is a bug in this PR, will fix

poettering commented on src/nspawn/nspawn-oci.c:
```diff
@@ -0,0 +1,2350 @@
+                        return -EINVAL;
+                }
+
+                assert
```
So, the json parser already filtered out these errors here, as we required that the sysctl object is actually an object. And an object has to have every second element as a string (because that's the fields key name). And every other second element can be something else, but we checked that it is a string a few lines above. Hence, when pulling out the two strings from `k` and `v` here we know for sure in both cases that it is a string. Or to say this differently: if the variant here is not an object, then json_dispatch() already has complained for us via json_log(). If the `k` is not a string then the parser should have complained even earlier already, also with json_log(), and if `v` is not a string we log with json_log() right here, so all grounds should be covered: we log in all bad cases, always friendly, and the assert_se() is only used to catch invalid expectations, i.e. if we cannot get a string for something that already was veryified to be a string.

poettering replied:
oh, and the oci spec says to refuse unknown fields, and says that the oci version at the top of the json data should be updated on extensions

poettering commented:
The promised PRs for the last mentioned issues are #11902 and #11901 btw.

poettering commented:
I force pushed a new version now. PTAL. It addresses all issues pointed out. (well, all but two, which are fixes in the two other PRs, see above).

poettering commented:
I rebased and added a fix for #11755 on top now.

keszybz commented:
Oh, I'll submit a PR with some follow-up cleanups shortly.

poettering commented on src/basic/capability-util.c:
```diff
@@ -47,6 +47,13 @@ unsigned long cap_last_cap(void) {
+                        if (p > 63) /* Safety for the future: if one day the kernel learns more than 64 caps,
+                                     * then we are in trouble (since we, as much userspace and kernel space
+                                     * store capability masks in uint64_t types. Let's hence protect
+                                     * ourselves against that and always cap at 63 for now. */
```
will add

poettering commented on src/nspawn/nspawn-mount.c:
```diff
@@ -815,11 +825,59 @@ static int mount_overlay(const char *dest, CustomMount *m) {
+
+        r = mount_verbose(m->graceful ? LOG_DEBUG : LOG_ERR, NULL, where, NULL, MS_BIND|MS_RDONLY|MS_REMOUNT, NULL);
+        if (r < 0)
+                return m->graceful ? 0 : r;
```
will add
